Libraries/MLXLLM/Models/GPTOSS.swift

```diff
@@ -0,0 +1,669 @@
+            return makeMask(L, offset)
+        }
+
+        let length = ((L + offset + 511) / 512) * 512
+        if self._previousMask == nil || self._previousMask!.dim(-1) < length
+            || self._previousMask!.dim(-2) != L
+        {
+            self._previousMask = makeMask(L, length - L)
+        }
+
+        return self._previousMask![.ellipsis, 0 ..< L + offset]
+    }
+
+    func getSlidingWindowMask(_ x: MLXArray, cache: KVCache?, windowSize: Int) -> MLXArray {
+        let L = x.dim(1)
+        var offset = cache?.offset ?? 0
+        offset = max(1, offset)
+
+        func makeMask(_ L: Int, _ offset: Int) -> MLXArray {
+            let zero = MLXArray(0, dtype: x.dtype)
+            let neginf = MLXArray(-Float.infinity, dtype: x.dtype)
+            var mask = createCausalMask(n: L, offset: offset - 1, windowSize: windowSize)
+            mask = MLX.where(mask, zero, neginf)
+            mask = mask.reshaped(1, 1, L, -1)
+            mask = tiled(mask, repetitions: [1, numAttentionHeads, 1, 1])
+            let sinks = tiled(sinks.reshaped(1, -1, 1, 1), repetitions: [1, 1, L, 1])
+            mask = concatenated([sinks, mask], axis: -1)
+            return mask
+        }
+
+        if training || L > 1 {
+            self._previousMask = nil
+            return makeMask(L, min(windowSize + 1, offset))
+        }
+
+        if self._previousMask == nil {
+            self._previousMask = makeMask(L, windowSize + 1)
+        }
+
+        return self._previousMask![.ellipsis, 0 ..< min(L + offset, windowSize + 1)]
+    }
+
+    func getMask(_ x: MLXArray, cache: KVCache?, windowSize: Int?) -> MLXArray {
+        if let windowSize {
+            return getSlidingWindowMask(x, cache: cache, windowSize: windowSize)
+        } else {
+            return getCausalMask(x, cache: cache)
+        }
+    }
+
+    public func callAsFunction(_ x: MLXArray, mask: MLXArray, cache: KVCache? = nil) -> MLXArray {
+        let (B, L) = (x.dim(0), x.dim(1))
+        let D = headDim
+        let Hk = numKeyValueHeads
+
+        var q = qProj(x).reshaped(B, L, -1, D).swappedAxes(1, 2)
+        var k = kProj(x).reshaped(B, L, -1, D).swappedAxes(1, 2)
+        var v = vProj(x).reshaped(B, L, -1, D).swappedAxes(1, 2)
+
+        if cache == nil || cache?.offset == 0 {
+            q = rope(q)
+            k = rope(k)
+
+            let zeros = MLXArray.zeros([B, Hk, 1, D]).asType(k.dtype)
+            k = concatenated([zeros, k], axis: 2)
+            v = concatenated([zeros, v], axis: 2)
+            if let cache {
+                (k, v) = cache.update(keys: k, values: v)
+            }
+        } else {
+            q = rope(q, offset: cache!.offset - 1)
+            k = rope(k, offset: cache!.offset - 1)
+            (k, v) = cache!.update(keys: k, values: v)
+        }
+
+        let vHat = MLXFast.scaledDotProductAttention(
+            queries: q, keys: k, values: v,
+            scale: smScale,
+            mask: mask
+        )
```

Comment:
Currently can't work. Still researching how to solve it.

```bash
MLX/ErrorHandler.swift:332: Fatal error: [broadcast_shapes] Shapes (1,64,1,130) and (1,64,1,129) cannot be broadcast. at /Users/john/Library/Developer/Xcode/DerivedData/mlx-swift-examples-aaenjefvhudawxerugkkmbixujtd/SourcePackages/checkouts/mlx-swift/Source/Cmlx/mlx-c/mlx/c/fast.cpp:398
```

<img width="1866" height="703" alt="Xcode 2025-08-07 23 25 11" src="https://github.com/user-attachments/assets/740cc1c3-56c9-4035-98ef-dfac1ca4673c" />